tests/linux/test_platform_linux.cpp

```diff
@@ -554,4 +557,185 @@ INSTANTIATE_TEST_SUITE_P(
+
+    auto output = multipass::platform::detail::find_os_release();
+    EXPECT_EQ(output->fileName(), expected_filename);
+}
+
+TEST_F(PlatformLinux, find_os_release_usr_lib)
+{
+    const auto expected_filename = QStringLiteral("/var/lib/snapd/hostfs/usr/lib/os-release");
+
+    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+
+    InSequence seq;
+    EXPECT_CALL(*mock_file_ops, open(Property(&QFile::fileName, Eq("/var/lib/snapd/hostfs/etc/os-release")),
+                                     QIODevice::ReadOnly | QIODevice::Text))
+        .WillOnce(Return(false));
+    EXPECT_CALL(*mock_file_ops,
+                open(Property(&QFile::fileName, Eq(expected_filename)), QIODevice::ReadOnly | QIODevice::Text))
+        .WillOnce(Return(true));
+    EXPECT_CALL(*mock_file_ops, open).Times(0); // no other open attempts
+
+    auto output = multipass::platform::detail::find_os_release();
+    EXPECT_EQ(output->fileName(), expected_filename);
+}
+
+TEST_F(PlatformLinux, read_os_release_from_file_not_found)
+{
+    const std::string expected = "unknown-unknown";
+
+    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+    EXPECT_CALL(*mock_file_ops, open(_, _)).Times(2).WillRepeatedly(Return(false));
+
+    auto output = multipass::platform::detail::read_os_release();
+
+    EXPECT_EQ(expected, output);
+}
+
+TEST_F(PlatformLinux, read_os_release_from_file)
+{
+    const std::string expected = "distribution_name-distribution_release";
+    const auto expected_filename = QStringLiteral("/var/lib/snapd/hostfs/etc/os-release");
+    const size_t os_rel_len = 13; // 12 lines of data + 1 for EOF.
+    QTextStream os_rel_file("NAME=\"distribution_name\"\n"
+                            "VERSION=\"21.04 (Hirsute Hippo)\"\n"
+                            "ID=ubuntu\n"
+                            "ID_LIKE=debian\n"
+                            "PRETTY_NAME=\"Ubuntu 21.04\"\n"
+                            "VERSION_ID=\"distribution_release\"\n"
+                            "HOME_URL=\"https://www.ubuntu.com/\"\n"
+                            "SUPPORT_URL=\"https://help.ubuntu.com/\"\n"
+                            "BUG_REPORT_URL=\"https://bugs.launchpad.net/ubuntu/\"\n"
+                            "PRIVACY_POLICY_URL=\"https://www.ubuntu.com/legal/terms-and-policies/privacy-policy\"\n"
+                            "VERSION_CODENAME=hirsute\n"
+                            "UBUNTU_CODENAME=hirsute\n");
+
+    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();
+
+    InSequence seq;
+    EXPECT_CALL(*mock_file_ops, open).WillOnce(Return(true));
+    EXPECT_CALL(*mock_file_ops, open).Times(0); // no other open attempts
+    EXPECT_CALL(*mock_file_ops, isOpen).Times(1).WillOnce(Return(true));
+    for (size_t idx = 0; idx < os_rel_len; ++idx)
+    {
+        EXPECT_CALL(*mock_file_ops, readLine).WillOnce(Return(os_rel_file.readLine())).RetiresOnSaturation();
+    }
```

Comment:
If you had `output` as a `QStringList` (as per suggestion above), you could use a range-for:

```
for (const auto& line : output)
  EXPECT_CALL(...).WillOnce(Return(line)).RetiresOnSaturation();
```

Or you can have a single expectation using an iterator and a lambda (since `ReturnRoundRobin` is not available in the gmock version we use). Something of the sort:

```
auto it = output.begin();
EXPECT_CALL(...).WillRepeatedly([&it, end = output.end()] {
    return it == end ? "" : *it++;
});

Reply:
> f you had `output` as a `QStringList` (as per suggestion above), you could use a range-for:

This makes things a lot cleaner and simpler, good suggestion - thanks!

Reply:
A bit of a conundrum as to why `read_line` seems to be called twice (over-saturated)
```c++
TEST_F(PlatformLinux, read_os_release_from_file)
{
    const std::string expected = "A-B";
    const QStringList& input = parse_os_release_single_char_fields.first;
    auto [mock_file_ops, guard] = mpt::MockFileOps::inject();

    InSequence seq;
    EXPECT_CALL(*mock_file_ops, open).WillOnce(Return(true));
    EXPECT_CALL(*mock_file_ops, is_open).WillOnce(Return(true));
    for (const QString& line : input)
    {
        EXPECT_CALL(*mock_file_ops, read_line).WillOnce(Return(line));
    }

    auto output = multipass::platform::detail::read_os_release();

    EXPECT_EQ(expected, output);
}
```
Setting it to retire on saturation does not fix the issue, and I do not feel it should be required seeing as how the `input` variable is updated at the loop and not in the mock call.

Reply:
Right, the code does `while (!line.isNull())`, so we need a null string at the end... I am trying something, I'll get back to you.